src/Raven.Client/ServerWide/Operations/ServerOperationExecutor.cs

```diff
@@ -10,22 +12,74 @@ namespace Raven.Client.ServerWide.Operations
 {
     public class ServerOperationExecutor : IDisposable
     {
+        private readonly ConcurrentDictionary<string, ServerOperationExecutor> _cache;
+
+        private readonly string _nodeTag;
+        private readonly DocumentStoreBase _store;
+
         private readonly ClusterRequestExecutor _requestExecutor;
+        private readonly ClusterRequestExecutor _initialRequestExecutor;
 
         public ServerOperationExecutor(DocumentStoreBase store)
+            : this(store, CreateRequestExecutor(store), initialRequestExecutor: null, new ConcurrentDictionary<string, ServerOperationExecutor>(StringComparer.OrdinalIgnoreCase), nodeTag: null)
+        {
+        }
+
+        private ServerOperationExecutor(DocumentStoreBase store, ClusterRequestExecutor requestExecutor, ClusterRequestExecutor initialRequestExecutor, ConcurrentDictionary<string, ServerOperationExecutor> cache, string nodeTag)
         {
-            _requestExecutor = store.Conventions.DisableTopologyUpdates
-                ? ClusterRequestExecutor.CreateForSingleNode(store.Urls[0], store.Certificate, store.Conventions)
-                : ClusterRequestExecutor.Create(store.Urls, store.Certificate, store.Conventions);
+            if (store == null)
+                throw new ArgumentNullException(nameof(store));
+            if (requestExecutor == null)
+                throw new ArgumentNullException(nameof(requestExecutor));
+
+            _store = store;
+            _requestExecutor = requestExecutor;
+            _initialRequestExecutor = initialRequestExecutor;
+            _nodeTag = nodeTag;
+            _cache = cache;
 
             store.RegisterEvents(_requestExecutor);
 
-            store.AfterDispose += (sender, args) => _requestExecutor.Dispose();
+            store.AfterDispose += (sender, args) => Dispose();
         }
 
         public void Dispose()
         {
-            _requestExecutor.Dispose();
+            _requestExecutor?.Dispose();
+
+            if (_nodeTag != null)
```

Comment:
Why would we want to do this?
Isn't it better to have a cache that will live as long as the parent live?

Reply:
We are holding main server operation executor for the whole lifetime of the DocumentStore. I'm assuming that vast amount of usages will be without a dispose, but when someone will dispose then I'm removing that from the cache to not hold request executor.

Reply:
The problem is that we may have two threads, one that dispose, one that doesn't.
You give them both the same instance, and it gets dispose under you.

We need to either:
* Give a different instance (and require disposal)
* Give same instance and forbid disposal

Reply:
Done